### PR TITLE
nixos/mediawiki: add nginx option

### PR DESCRIPTION
**Pull request by Mic92**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
@ofborg test mediawiki

**Review by mweinelt — APPROVED** [review]

**Comment by mweinelt** [comment]
This leads to paths like `https://wiki.example.com/index.php/Main_page`, where as we should probably follow what Wikipedia does, which is `/wiki/Main_page`. Pretty URLs and such.

**Comment by fleaz** [comment]
Finaly nginx support. Thanks \o/

I just deployed this to an existing Mediawiki of me, and it basically works pretty well but I think I found a little problem:

When you provide `services.mediawiki.uploadsDir` like e.g. `/srv/mediawiki` you get a location block for the images which looks like this:

```
location /w/images/ {
    alias /srv/mediawiki;
}
```

This will result in all images in the wiki being broken because nginx builds a broken filepaths and you get an error in the log like this `open() "/srv/mediawikid/d7/Header2023.png" failed`

To make images work you need to provide `services.mediawiki.uploadsDir` with a trailing slash like e.g. `/srv/mediawiki/`. With this little change the module works perfectly from what I have seen.

Maybe a little helper like this could help us here?
```
withTrailingSlash = str: if hasSuffix "/" str then str else "${str}/";
```
(taken from services/web-apps/jirafeau.nix)

**Comment by Mic92** [comment]
Added the helper.

**Comment by McSinyx** [comment]
Some skin resources don't seem to show up [for us](https://nhanh.loang.net), e.g. `/w/skins/Vector/resources/common/images/search.svg` (404), could anyone else reproduce?

**Comment by Mic92** [comment]
Seems to also do a 404 here: https://nixos-wiki.thalheim.io/w/skins/Vector/resources/common/images/search.svg
You can debug these types issues by using `strace -p <nginx-pid> -e openat,stat`

File: nixos/modules/services/web-apps/mediawiki.nix [hunk 1]
```diff
@@ -503,6 +533,62 @@ in
...
+          '';
+          "/w/images/".alias = withTrailingSlash cfg.uploadsDir;
+          # Deny access to deleted images folder
+          "/w/images/deleted".extraConfig = ''
+            deny all;
+          '';
+          # MediaWiki assets (usually images)
+          "~ ^/w/resources/(assets|lib|src)" = {
+            tryFiles = "$uri =404";
+            extraConfig = ''
```
Comment (McSinyx) [inline nixos/modules/services/web-apps/mediawiki.nix:557]: This should have been

```suggestion
            extraConfig = ''
              rewrite ^/w(/.*) $1 break;
```

File: nixos/modules/services/web-apps/mediawiki.nix [hunk 2]
```diff
@@ -503,6 +533,62 @@ in
...
+              add_header Cache-Control "public";
+              expires 7d;
+            '';
+          };
+          # Assets, scripts and styles from skins and extensions
+          "~ ^/w/(skins|extensions)/.+\\.(css|js|gif|jpg|jpeg|png|svg|wasm|ttf|woff|woff2)$" = {
+            tryFiles = "$uri =404";
+            extraConfig = ''
```
Comment (McSinyx) [inline nixos/modules/services/web-apps/mediawiki.nix:565]: Same here.

**Review by McSinyx — COMMENTED** [review]
REST API also seems to be broken.

File: nixos/modules/services/web-apps/mediawiki.nix (hunk 1 above)
Comment (Mic92) [reply nixos/modules/services/web-apps/mediawiki.nix:557]: Can you make a pull request?

Comment (Mic92) [reply nixos/modules/services/web-apps/mediawiki.nix:557]: Or if not, how can I test these changes?

Comment (McSinyx) [reply nixos/modules/services/web-apps/mediawiki.nix:557]: [These](https://trong.loang.net/nixos-conf/commit/?id=5f1653842dd91fbeff94472f6e3848f12492fd9f) are deployed on our wiki instance, e.g. [/w/skins/Vector/resources/common/images/link-external-small-ltr-progressive.svg](https://nhanh.loang.net/w/skins/Vector/resources/common/images/link-external-small-ltr-progressive.svg) now loads.  The same fix doesn't work for `/w/rest.php` though so I'm hesitant to submit the incomplete patch.

**Comment by Mic92** [comment]
@McSinyx some follow up fixes: https://github.com/NixOS/nixpkgs/pull/265506

File: nixos/modules/services/web-apps/mediawiki.nix [hunk 3]
```diff
@@ -209,8 +220,14 @@ in
...
+          if cfg.webserver == "apache" then
             "${if cfg.httpd.virtualHost.addSSL || cfg.httpd.virtualHost.forceSSL || cfg.httpd.virtualHost.onlySSL then "https" else "http"}://${cfg.httpd.virtualHost.hostName}"
+          else if cfg.webserver == "nginx" then
+            let
+              hasSSL = host: host.forceSSL || host.addSSL;
+            in
+            "${if hasSSL config.services.nginx.virtualHosts.${cfg.nginx.hostName} then "https" else "http"}://${cfg.nginx.hostName}"
           else
             "http://localhost";
         defaultText = literalExpression ''
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mediawiki.nix:233]: You forgot to update the defaulText here

Comment (mweinelt) [reply nixos/modules/services/web-apps/mediawiki.nix:233]: Feel free to send a PR.

Comment (SuperSandro2000) [reply nixos/modules/services/web-apps/mediawiki.nix:233]: #269423

**Pull request merged**